dkg-primitives/src/rounds/keygen.rs

```diff
@@ -0,0 +1,288 @@
+
+			match keygen.proceed() {
+				Ok(_) => {
+					trace!(target: "dkg", "🕸️  after: {:?}", keygen);
+				},
+				Err(err) => {
+					match err {
+						gg20_keygen::Error::ProceedRound(proceed_err) => match proceed_err {
+							gg20_keygen::ProceedError::Round2VerifyCommitments(err_type) =>
+								return Err(DKGError::KeygenMisbehaviour {
+									bad_actors: vec_usize_to_u16(err_type.bad_actors),
+								}),
+							gg20_keygen::ProceedError::Round3VerifyVssConstruct(err_type) =>
+								return Err(DKGError::KeygenMisbehaviour {
+									bad_actors: vec_usize_to_u16(err_type.bad_actors),
+								}),
+							gg20_keygen::ProceedError::Round4VerifyDLogProof(err_type) =>
+								return Err(DKGError::KeygenMisbehaviour {
+									bad_actors: vec_usize_to_u16(err_type.bad_actors),
+								}),
+						},
+						_ => return Err(DKGError::GenericError { reason: err.to_string() }),
+					};
+				},
+			}
+		}
+
+		let (_, blame_vec) = keygen.round_blame();
+
+		if keygen.is_finished() {
+			Ok(true)
+		} else {
+			if at - self.started_at > KEYGEN_TIMEOUT.into() {
+				if !blame_vec.is_empty() {
+					return Err(DKGError::KeygenTimeout { bad_actors: blame_vec })
+				} else {
+					// Should never happen
+					warn!(target: "dkg", "🕸️  Keygen timeout reached, but no missing parties found", );
+				}
+			}
+			Ok(false)
+		}
+	}
+
+	/// Get outgoing messages
+
+	fn get_outgoing(&mut self) -> Vec<DKGKeygenMessage> {
+		trace!(target: "dkg", "🕸️  Getting outgoing keygen messages");
+
+		let keygen = &mut self.keygen;
+
+		if !keygen.message_queue().is_empty() {
+			trace!(target: "dkg", "🕸️  Outgoing messages, queue len: {}", keygen.message_queue().len());
+
+			let keygen_set_id = self.params.keygen_set_id;
+
+			let enc_messages = keygen
+				.message_queue()
+				.into_iter()
+				.map(|m| {
+					trace!(target: "dkg", "🕸️  MPC protocol message {:?}", m);
+					let serialized = serde_json::to_string(&m).unwrap();
+					return DKGKeygenMessage { keygen_set_id, keygen_msg: serialized.into_bytes() }
+				})
+				.collect::<Vec<DKGKeygenMessage>>();
+
+			keygen.message_queue().clear();
+			return enc_messages
+		}
+
+		vec![]
+	}
+
+	/// Handle incoming messages
+
+	fn handle_incoming(&mut self, data: DKGKeygenMessage, at: C) -> Result<(), DKGError> {
+		if data.keygen_set_id != self.params.keygen_set_id {
+			return Err(DKGError::GenericError { reason: "Keygen set ids do not match".to_string() })
+		}
+
+		let keygen = &mut self.keygen;
+
+		trace!(target: "dkg", "🕸️  Handle incoming keygen message");
+		if data.keygen_msg.is_empty() {
+			warn!(
+				target: "dkg", "🕸️  Got empty message");
+			return Ok(())
+		}
+
+		let msg: Msg<ProtocolMessage> = match serde_json::from_slice(&data.keygen_msg) {
+			Ok(msg) => msg,
+			Err(err) => {
+				error!(target: "dkg", "🕸️  Error deserializing msg: {:?}", err);
+				return Err(DKGError::GenericError {
+					reason: "Error deserializing keygen msg".to_string(),
```

Comment:
we could use `format!("Error deserializing keygen msg, reason: {}", e),` so we get the actual error message too.